lib/campfire_social_hour/zoom_credential_agent.ex

```diff
@@ -0,0 +1,68 @@
+defmodule CampfireSocialHour.ZoomCredentialAgent do
+  @moduledoc """
+  Agent responsible for maintaining zoom client credential state
+  """
+  use Agent
+  require Logger
+
+  @type state ::
+          nil
+          | %{
+              access_token: String.t(),
+              expires_at: integer
+            }
+
+  def start_link(_arg),
+    do: Agent.start_link(fn -> nil end, name: __MODULE__)
+
+  @spec access_token :: String.t()
+  def access_token(), do: Agent.get_and_update(__MODULE__, &get_and_update_token/1)
+
+  @spec get_token() :: state
+  defp get_token() do
+    debug("Getting Access Token")
+
+    case zoom_api().get_access_token() do
+      {:ok, %{access_token: access_token, expires_in: expires_in}} ->
+        %{access_token: access_token, expires_at: now() + expires_in}
+
+      :error ->
+        nil
+    end
+  end
+
+  @spec token_to_update_tuple(nil | map) :: {nil | String.t(), state}
+  defp token_to_update_tuple(token) do
+    case token do
+      %{access_token: access_token, expires_at: _} = new_state -> {access_token, new_state}
+      nil = new_state -> {nil, new_state}
+    end
+  end
+
+  @spec get_and_update_token(state) :: {String.t(), state}
+  defp get_and_update_token(_state = nil) do
+    debug("Token not set, Fetching")
+
+    get_token()
+    |> token_to_update_tuple()
+  end
+
+  defp get_and_update_token(state = %{expires_at: expires_at, access_token: access_token}) do
+    if now() + 1 > expires_at do
+      debug("Token Expired, Refreshing")
+
+      get_token()
+      |> token_to_update_tuple()
```

Comment:
A downside of this design is the caller would have to wait for the new token and thus incur latency. And if we have N concurrent callers, we'd request the token N times.

To be fair this really depends on the app, if we don't have a lot of traffic it might be fine as is.

But for completeness, of course when we fetch the token for the first time we need to wait for that to finish, but we can avoid subsequent waits by prefetching the token sometime before it would have expired. If you'd like to explore that solution, here's a rough sketch:

```elixir
defmodule ZoomToken do
  use GenServer

  def start_link(options \\ []) do
    GenServer.start_link(__MODULE__, nil, name: options[:name] || __MODULE__)
  end

  def fetch(name \\ __MODULE__, timeout \\ 5000) do
    GenServer.call(name, :fetch, timeout)
  end

  @impl true
  def init(nil) do
    # to improve cache hits, we can prefetch the token. If we request it right here,
    # prefetching will be synchronous, that is, the supervision tree waits for this
    # init/1 to finish. We can make it asynchronous too, by sending ourselves
    # `send(self(), :refresh)` or using `handle_continue`.
    # About sync prefetch, be careful around error handling,
    # if requesting token will immediately keep failing, this genserver will
    # keep restarting, and it will crash the app.
    {:ok, %{token: nil}}
  end

  @impl true
  def handle_call(:fetch, _from, state) do
    if state.token do
      {:reply, state.token, state}
    else
      token = request_token()
      schedule_refresh(token)
      {:reply, token, %{state | token: token}}
    end
  end

  @impl true
  def handle_info(:refresh, state) do
    token = request_token()
    schedule_refresh(token)
    {:noreply, %{state | token: token}}
  end

  defp request_token do
    # If this crashes, our genserver would crash too but it's ok
    # because app supervisor will restart it. If we get an http failure, might be a good
    # idea to retry a few times with some backoff.
    {:ok, token} = zoom_api().get_access_token()
    token
  end

  defp schedule_refresh(token) do
    # 5 minutes
    time_before_expire = 5 * 60

    Process.send_after(self(), :refresh, abs(token.expires_in - time_before_expire) * 1000)
  end
end
```

You can learn more about this in our [Goth Redesign](https://dashbit.co/blog/goth-redesign) blog post.